apisec/internal/timed/table.go

```diff
@@ -0,0 +1,190 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023-present Datadog, Inc.
+
+package timed
+
+import (
+	"container/heap"
+	"sync/atomic"
+
+	"github.com/DataDog/appsec-internal-go/apisec/internal/config"
+)
+
+type (
+	// table is a simple open-addressing hash table implementation that uses a
+	// fixed array of items.
+	table struct {
+		// entries is the set of items contained in the table. The last entry is
+		// reserved for cases where all slots are taken before a rebuild is
+		// complete.
+		entries [capacity + 1]entry
+		// count is the number of items currently stored in the table.
+		count atomic.Int32
+	}
+
+	// entry is a single item in the open-addressing hash table.
+	entry struct {
+		// Key is the Key of the entry. A zero Key indicates that the entry is
+		// currently free.
+		Key atomic.Uint64
+		// Data is the Data associated with the entry.
+		Data atomicEntryData
+	}
+
+	// atomicEntryData is an atomic version of [entryData].
+	atomicEntryData atomic.Uint64
+	// entryData is a 64-bit value that represents the last time an entry was
+	// accessed paired together with the last time this value was sampled.
+	entryData uint64
+
+	// copiableEntry is a copy-able version of entryData, which is used for
+	// sorting entries by recency using a heap when re-building the table.
+	copiableEntry struct {
+		// Key is the Key of the entry.
+		Key uint64
+		// Data is the Data associated with the entry.
+		Data entryData
+	}
+	// sortedEntries is a [heap.Interface] used to sort [copiableEntry] when
+	// rebuilding the table in order to evict older items.
+	sortedEntries []copiableEntry
+)
+
+// FindEntry locates the correct entry for use in the table. If an entry already
+// exists for the given key, it is returned with true. If not, the first blank
+// entry is returned with false.
+func (t *table) FindEntry(key uint64) (*entry, bool) {
+	origIdx := key % capacity
+	idx := origIdx
+
+	for {
+		entry := &t.entries[idx]
+		switch entry.Key.Load() {
+		case key:
+			// This entry is the one we're looking for...
+			return entry, true
+
+		case 0:
+			// This entry is currently blank, so we can use it!
+			return entry, false
+		}
```

Comment:
```suggestion
		if currKey := entry.Key.Load(); currKey == 0 || currKey == key {
			return entry, currKey == key
		}
```
This is very much less ambiguous I believe because we naturally tend to think that a switch statement cover all cases and this one does not voluntarily